edx/edx-enterprise
# Bulk upload users

**Description:** This PR adds ability to bulk link users to enterprise customer.

**JIRA:** [ENT-62](https://openedx.atlassian.net/browse/ENT-62)

**Dependencies:** cross dependent on https://github.com/edx/edx-platform/pull/14003

**Testing instructions:**

1. Linking users by username or email was significantly affected - rerun #8 tests.
2. Prepare CSV file (*Yes, so far there are just one column, so it is more like "one record per line", but CSV is a superset of it anyway*).
2.1. First line should contain column header `email_or_username`
2.2. Each subsequent line should contain existing user username or an email.
2.3. Don't do any intentional mistakes to test how it will behave yet - we'll get to it later. Make sure the list does not contain users already linked to ECs (even the current one) or invalid emails.
3. Go to "Manage Learners" view in Django admin (admin -> Enterprise Customers -> choose one -> "Manage Learners" in top-right corner)
4. Test "Manage Learners" form validation:
4.1. Submit empty form - should complain that either username/email or bulk upload CSV must be present.
4.2. Submit both username/email and bulk upload CSV - should complain about both being present.
4.3. Submit invalid email only - should complain that email is invalid.
5. Try uploading the file created in step 2. Existing users should be linked to EC, unknown emails should end up in Pending EC Users
6. Try uploading the same file again (yes, don't modify the file, don't unlink anything). *Expected result:* Form explodes with a ton of errors. Errors should make sense so that non-technical person could understand what needs to be fixed.
7. Unlink (some) records, change the CSV (introduce invalid emails, already linked users, already existing pending links, etc.). Upload the file again. *Expected result:* form errors are shown, errors still make sense. Note that "valid" records are still supposed to be added to corresponding lists: valid emails with not used in pending link - in Pending Linked Users and usernames of existing users not linked to any EC - in Linked users.

**Reviewers:**
- [x] @asadiqbal08 and/or @saleem-latif 

**Merge checklist:**

**Post merge:**
